src/server/lib/okta/register.ts

```diff
@@ -109,13 +99,62 @@ export const sendRegistrationEmailByUserState = async (
     case Status.ACTIVE: {
       /* Given I'm an ACTIVE user
        *    When I try to register
-       *    Then I should be sent an email with a link to the reset
-       *    password form (with no token)
+       *    Given I don't have a validated email
+       *        Given I have a password set
+       *            Then I should be sent an email with a reset password token
+       *        Given I do NOT have a password set
+       *            Then I should have my password dangerously reset, and then
+       *            sent an email with a reset password token
+       *    Given I have a validated email
+       *        Then I should be sent an email with NO reset password token
        *    And my status should remain ACTIVE
        */
-      await sendAccountExistsEmail({
-        to: user.profile.email,
-      });
+
+      const doesNotHavePassword = !user.credentials.password;
+
+      const groups = await getUserGroups(id);
+      // check if the user has their email validated based on group membership
+      const emailValidated = groups.some(
+        (group) => group.profile.name === 'GuardianUser-EmailValidated',
+      );
```

Comment:
why do we get the user groups here ? can't we just check the `emailValidated` boolean flag on the user ? instead of making an additional api call ?

Reply:
We use the groups as our source of truth, because there could be a situation where a user is removed from a group by an admin or new rule, but still has the flag set on their profile. And because the flags don't actually mean anything to Okta, but groups do (in the sense of allowing users to do certain things)

Reply:
Ah I found the actual reason in the description for this commit: https://github.com/guardian/gateway/commit/255ccbebd2dc78824836caff9428dd7401418ea3

Reply:
> We use the groups api to get the group membership, rather than checking the email validated field because
> social users do not have the email validated flag set until after they login for the first time (but are in the email validated group)

Reply:
So there as edge cases where the user might have the group but not the flag, hence the groups should be the source of truth for this!

Reply:
A question you might have is then how do we reconcile these two things?

Well, when the user finishes logging in, and in the OAuth callback route on gateway we check for this, and update the user profile if appropriate:

https://github.com/guardian/gateway/blob/main/src/server/routes/oauth.ts#L162-L182

Reply:
But it's still best to use the groups rather than the field to determine if the user is in the correct state!

Reply:
would this be a problem in other cases ? I mean can we safely rely on the flag everywhere else ?

Reply:
wouldn't this violate the group rule we set in okta ? Or does the rule only enforce that people in the group have the flag set to true but allows adding people to the group with the flag to false ? 🤔

Reply:
I think it's probably best to always rely on the group as a matter of course...